pennylane/transforms/optimization/pattern_matching.py

```diff
@@ -0,0 +1,1638 @@
+# Copyright 2021-2022 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Transform finding all maximal matches of a pattern in a quantum circuit and optimizing the circuit by
+substitution."""
+
+import itertools
+import copy
+from collections import OrderedDict
+
+import numpy as np
+
+import pennylane as qml
+from pennylane import apply
+from pennylane.transforms import qfunc_transform
+from pennylane.transforms.commutation_dag import commutation_dag
+from pennylane.wires import Wires
+from pennylane.ops.qubit.attributes import (
+    symmetric_over_all_wires,
+)
+
+
+@qfunc_transform
+def pattern_matching_optimization(tape, pattern_tapes, custom_quantum_cost=None):
+    r"""Quantum function transform to optimize a circuit given a list of patterns (templates).
+
+    Args:
+        qfunc (function): A quantum function to be optimized.
+        pattern_tapes(list(.QuantumTape)): List of quantum tapes that implements the identity.
+        custom_quantum_cost (dict): Optional, quantum cost that overrides the default cost dictionnary.
+
+    Returns:
+        function: the transformed quantum function
+
+    Raises:
+        QuantumFunctionError: The pattern provided is not a valid QuantumTape or the pattern contains measurements or
+            the pattern does not implement identity or the circuit has less qubits than the pattern.
+
+    **Example**
+
+    Consider the following quantum circuit to be optimized
+
+    .. code-block:: python
+
+        def circuit():
```

Comment:
I wonder if it would be better to start with a blatantly obvious example, where a human can see the pattern without having to stare at it for too long, and then show the more complicated template example.

Like the template is something like `[qml.S(0), qml.S(0), qml.PauliZ(0)]`.